examples/petsc/dmswarm.c

```diff
@@ -0,0 +1,410 @@
+      PetscCall(PetscFECreateTabulation(fe, 1, num_points_in_cell, coords_points_ref, 1, &tabulation));
+      PetscCall(DMPlexComputeCellGeometryFEM(dm_mesh, cell, NULL, v, J, invJ, &detJ));
+      PetscCall(DMPlexVecGetClosure(dm_mesh, NULL, U_loc, cell, NULL, &u_cell));
+      for (PetscInt p = 0; p < num_points_in_cell; p++) {
+        PetscScalar x[dim];
+
+        PetscCall(PetscFEGetQuadrature(fe, &quadrature));
+        PetscCall(PetscFECreateCellGeometry(fe, quadrature, &fe_geometry));
+        PetscCall(PetscFEInterpolateAtPoints_Static(fe, tabulation, u_cell, &fe_geometry, p, &u_points_array[points[p]]));
+        PetscCall(PetscFEDestroyCellGeometry(fe, &fe_geometry));
+        for (PetscInt d = 0; d < dim; d++) x[d] = coords_points_true[p * dim + d];
+        PetscCheck(PetscAbs(EvalU(dim, x) - u_points_array[points[p]]) > 1E-4, comm, PETSC_ERR_USER, "Incorrect interpolated value from PETSc");
+      }
+
+      // ---- Cleanup
+      PetscCall(DMPlexVecRestoreClosure(dm_mesh, NULL, U_loc, cell, NULL, &u_cell));
+      PetscCall(DMRestoreWorkArray(dm_mesh, num_points_in_cell * dim, MPIU_REAL, &coords_points_true));
+      PetscCall(DMRestoreWorkArray(dm_mesh, num_points_in_cell * dim, MPIU_REAL, &coords_points_ref));
+      PetscCall(PetscTabulationDestroy(&tabulation));
+      PetscCall(PetscFree(points));
+    }
+
+    // -- Cleanup
+    PetscCall(DMSwarmRestoreField(dm_swarm, "DMSwarmPIC_coor", NULL, NULL, (void **)&coord_points));
+    PetscCall(DMSwarmRestoreField(dm_swarm, "u", NULL, NULL, (void **)&u_points_array));
+    PetscCall(DMSwarmSortRestoreAccess(dm_swarm));
+    PetscCall(DMRestoreLocalVector(dm_mesh, &U_loc));
+  }
+
+  // Interpolate from mesh to points via libCEED
+  {
+    IS                 is_points;
+    Vec                U_loc, ref_coords;
+    PetscInt           num_points_local;
+    const PetscInt    *all_points;
+    PetscInt           cell_start, cell_end;
+    PetscScalar       *u_points_array;
+    const PetscScalar *coords_points_ref;
+    CeedVector         u_evec, u_cell;
+
+    // -- Get mesh values
+    PetscCall(DMGetLocalVector(dm_mesh, &U_loc));
+    PetscCall(DMGlobalToLocal(dm_mesh, U_mesh, INSERT_VALUES, U_loc));
+    {
+      const PetscScalar *u_array;
+      CeedVector         u_lvec;
+
+      PetscCall(VecGetArrayRead(U_loc, &u_array));
+      CeedElemRestrictionCreateVector(restriction_mesh_u, &u_lvec, &u_evec);
+      CeedVectorSetArray(u_lvec, CEED_USE_POINTER, CEED_MEM_HOST, (CeedScalar *)u_array);
+      CeedElemRestrictionApply(restriction_mesh_u, CEED_NOTRANSPOSE, u_lvec, u_evec, CEED_REQUEST_IMMEDIATE);
+      CeedVectorTakeArray(u_lvec, CEED_MEM_HOST, (CeedScalar **)&u_array);
+      PetscCall(VecRestoreArrayRead(U_loc, &u_array));
+      CeedVectorDestroy(&u_lvec);
+    }
+    {
+      CeedInt P;
+
+      CeedBasisGetNumNodes(basis_mesh_u, &P);
+      CeedVectorCreate(ceed, P, &u_cell);
+    }
```

Comment:
We do this internally in the CPU backends a lot. It might be nice to have a utility to make a single element evec or a 8 element batch evec?